File: src/Chauffeur.Deliverables/HelpDeliverable.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using Umbraco.Core;
using Umbraco.Core.Composing;

namespace Chauffeur.Deliverables
{
    [DeliverableName("help")]
    [DeliverableAlias("h")]
    [DeliverableAlias("?")]
    public class HelpDeliverable : Deliverable, IProvideDirections
    {
        private readonly IFactory container;

        public HelpDeliverable(TextReader reader, TextWriter writer, IFactory container)
            : base(reader, writer)
        {
            this.container = container;
        }

        public override async Task<DeliverableResponse> Run(string command, string[] args)
        {
            if (args.Any())
                await Print(args[0]);
            else
                await PrintAll();

            return DeliverableResponse.Continue;
        }

        private async Task Print(string command)
        {
            if (container.GetInstance<Deliverable>() is IProvideDirections deliverable)
            {
                await deliverable.Directions();
                return;
            }

            await Out.WriteLineAsync(
                $"The deliverable '{command}' doesn't implement help, you best contact the author"
            );
        }

        private async Task PrintAll()
        {
            await Out.WriteLineAsync("The following deliverables are loaded. Use `help <deliverable>` for detailed help");

            var toWrite = new List<string>();
            var deliverables = container.GetAllInstances<Deliverable>().Distinct(new DeliverableComparer());

            foreach (var deliverable in deliverables)
            {
                var type = deliverable.GetType();
                var name = type.GetCustomAttribute<DeliverableNameAttribute>(false).Name;
                var aliases = type.GetCustomAttributes<DeliverableAliasAttribute>(false).Select(a => a.Alias);

                toWrite.Add(
                    string.Format(
                        "{0}{1}",
                        name,
                        aliases.Any() ? string.Format(" (aliases: {0})", string.Join(", ", aliases)) : string.Empty
                    )
                );
            }

            await Task.WhenAll(toWrite.OrderBy(s => s).Select(Out.WriteLineAsync).ToArray());
        }

        public async Task Directions()
        {
            await Out.WriteLineAsync("help");
            await Out.WriteLineAsync("\taliases: h, ?");
            await Out.WriteLineAsync("\tUse `help` to display system help");
            await Out.WriteLineAsync("\tUse `help <Deliverable>` to display help for a deliverable");
        }

        class DeliverableComparer : IEqualityComparer<Deliverable>
        {
            public bool Equals(Deliverable x, Deliverable y)
            {
                return x.GetType() == y.GetType();
            }

            public int GetHashCode(Deliverable obj)
            {
                return obj.GetType().GetHashCode();
            }
        }
    }
}

File: src/Chauffeur/Deliverables/HelpDeliverable.fs
namespace Chauffeur.Deliverables

open Chauffeur
open Chauffeur.Components
open Umbraco.Core
open Umbraco.Core.Composing
open FSharp.Control.Tasks.V2
open System.Threading.Tasks

[<DeliverableName("help")>]
[<DeliverableAlias("h")>]
[<DeliverableAlias("?")>]
type HelpDeliverable(reader, writer, container : IFactory) =
    inherit Deliverable(reader, writer)

    let printDeliverable name =
        let resolver = container.GetInstance<DeliverableResolver>()
        match resolver.Resolve name with
        | Some deliverable ->
            task {
                match box deliverable with
                | :? IProvideDirections as hd ->
                    let! _ = hd.Directions()
                    ignore()
                | _ ->
                    let msg = sprintf "The deliverable '%s' doesn't implement help, you best contact the author" name
                    do! writer.WriteLineAsync msg

                return DeliverableResponse.Continue
            }
        | None ->
            task {
                let msg = sprintf "The deliverable '%s' doesn't implement help, you best contact the author" name
                do! writer.WriteLineAsync msg
                return DeliverableResponse.Continue
            }

    let printAll() =
        let metadata d =
            let type' = d.GetType()
            let name = type'.GetCustomAttribute<DeliverableNameAttribute>(false).Name
            let aliases = type'.GetCustomAttributes<DeliverableAliasAttribute>(false) |> Seq.map (fun attr -> attr.Alias)

            (name, aliases)

        task {
            do! writer.WriteLineAsync "The following deliverables are loaded. Use `help <deliverable>` for detailed help."

            let deliverables = container.GetAllInstances<Deliverable>()

            let tasks = deliverables
                        |> Seq.map(metadata)
                        |> Seq.map(fun (name, aliases) ->
                                    let msg = match aliases |> Seq.length with
                                              | 0 -> name
                                              | _ ->
                                                sprintf "%s (aliases: %s)" name (aliases |> String.concat(", "))

                                    writer.WriteLineAsync(msg))

            do! Task.WhenAll(tasks)

            return DeliverableResponse.Continue
        }

    override __.Run _ args =
        task {
            match args |> Array.toList with
            | name :: _ ->
                return! printDeliverable name
            | [] ->
                return! printAll()
        }

    interface IProvideDirections with
        member __.Directions() =
            task {
                do! writer.WriteLineAsync("help")
                do! writer.WriteLineAsync("\taliases: h, ?")
                do! writer.WriteLineAsync("\tUse `help` to display system help")
                do! writer.WriteLineAsync("\tUse `help <Deliverable>` to display help for a deliverable")

                return true
            }